src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -76,12 +83,22 @@ public void completed(Integer result, Void v) {
                                 return;
                             }
                             byte[] data = buffer.array();
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel");
+                            }
                             callback.onGetComplete(entry.toCacheEntry(data));
                         }
 
                         @Override
                         public void failed(Throwable exc, Void v) {
                             VolleyLog.e(exc, "Failed to read file %s", file.getAbsolutePath());
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel");
+                            }
                             callback.onGetComplete(null);
                         }
                     });
```

Comment:
Should we close the afc in the catch below as well? If we failed to read the file, and respond to the callback, I don't expect we still need the channel to be open (assuming we were able to open it)

Reply:
It doesn't let me add it there, since the channel is defined in the try block. And I can't declare the channel outside of it, since it needs to be final. Is there any way around this? I think we should be closing it, since IOExceptions can be thrown between opening the channel and performing the read/write.

Reply:
Strictly speaking the final catch(IOE) should close the channel if it is open. To better isolate the different failures, I would add a try/finally after opening the channel. Where the finally block closes the channel. Then outer try/catch can just report the error. (See pattern in DiskBasedCache.)

OTOH, the afc.read can throw IOE (NonReadableChannelException) only if channel is not open for reading, so you might never need to close in this exceptional case, but comments are needed to explain, and may also be needed to disable static checks.

Reply:
I tried it with a finally block, but with an AsynchronousFileChannel, it will get closed before finishing the read/write, and then cause an error. And there can be an IOException thrown while writing the CacheHeader (UnsupportedEncodingException), so that would be a case where the channel wouldn't actually be closed. I think closing the channel in the catch block may be the only way to do it.

Reply:
@sphill99  I see what you mean now